<launch>

    <!-- Set to your sensor's range -->
    <arg name="sensor_range" default="1.0"/>
    <arg name="output_client" default="screen" />
    <arg name="output_server" default="screen" />
    <arg name="delay_time" default="0.0" />
    <arg name="method" default="information_gain" />
    <arg name="available_time" default="600.0" />
    <arg name="inf_obstacle" default="10.0" />
    <arg name="inf_obstacle_unexplored" default="20.0" />
    <arg name="inf_object" default="100.0" />
    <arg name="inf_object_unexplored" default="30.0" />
    <arg name="inf_unexplored" default="1.0" />

    <node pkg="frontier_exploration" type="explore_client" name="explore_client" output="$(arg output_client)"/>

    <node pkg="frontier_exploration" type="explore_server" name="explore_server" output="$(arg output_server)" required="true">

        <param name="exploration_delay_time" type="double" value="$(arg delay_time)" />

        <!--<remap from="move_base/goal" to="wobtata"/>-->
        <param name="frequency" type="double" value="2.0"/>
        <param name="goal_aliasing" type="double" value="$(arg sensor_range)"/>
        <param name="available_time" type="double" value="$(arg available_time)"/>
        <param name="method" type="str" value="$(arg method)"/>
        <param name="inf_obstacle" type="double" value="$(arg inf_obstacle)"/>
        <param name="inf_obstacle_unexplored" type="double" value="$(arg inf_obstacle_unexplored)"/>
        <param name="inf_object" type="double" value="$(arg inf_object)"/>
        <param name="inf_object_unexplored" type="double" value="$(arg inf_object_unexplored)"/>
        <param name="inf_unexplored" type="double" value="$(arg inf_unexplored)"/>

        #All standard costmap_2d parameters as in move_base, other than BoundedExploreLayer
        <rosparam ns="explore_costmap" subst_value="true">
            footprint: [[-0.17, -0.20], [-0.20, -0.10], [-0.225, 0.0], [-0.20, 0.10], [-0.17, 0.20], [0.10, 0.20], [0.17, 0.10], [0.225, 0.0], [0.17, -0.10], [0.10, -0.20]]
            robot_radius: 0.20

            transform_tolerance: 0.5
            update_frequency: 5.0
            publish_frequency: 5.0

            #must match incoming static map
            global_frame: map
            robot_base_frame: base_link
            resolution: 0.05

            rolling_window: false
            track_unknown_space: true

            plugins:

                - {name: static,           type: "costmap_2d::StaticLayer"}
                - {name: explore_boundary, type: "frontier_exploration::BoundedExploreLayer"}
                #Can disable sensor layer if gmapping is fast enough to update scans
                - {name: sensor,           type: "costmap_2d::ObstacleLayer"}
                - {name: inflation,        type: "costmap_2d::InflationLayer"}

            static:
                #Can pull data from gmapping, map_server or a non-rolling costmap
                map_topic: /map
                # map_topic: move_base/global_costmap/costmap
                subscribe_to_updates: true

            explore_boundary:
                resize_to_boundary: false
                frontier_travel_point: middle
                #set to false for gmapping, true if re-exploring a known area
                explore_clear_space: false
                method: $(arg method)
                circle_radius: 3.0
                horizontal_fov: 60.0

            sensor:
                observation_sources: laser
                laser: {data_type: LaserScan, clearing: true, marking: true, topic: scan, inf_is_valid: true, raytrace_range: $(arg sensor_range), obstacle_range: $(arg sensor_range)}

            inflation:
                inflation_radius: 0.2
                
      </rosparam>
    </node>

</launch>
